hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieCDCLogger.java

```diff
@@ -0,0 +1,263 @@
+      this.cdcSupplementalLoggingMode = HoodieCDCSupplementalLoggingMode.parse(
+          config.getStringOrDefault(HoodieTableConfig.CDC_SUPPLEMENTAL_LOGGING_MODE));
+
+      if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE_AFTER)) {
+        this.cdcSchema = HoodieCDCUtils.CDC_SCHEMA;
+      } else if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE)) {
+        this.cdcSchema = HoodieCDCUtils.CDC_SCHEMA_OP_RECORDKEY_BEFORE;
+      } else {
+        this.cdcSchema = HoodieCDCUtils.CDC_SCHEMA_OP_AND_RECORDKEY;
+      }
+
+      this.cdcData = new ExternalSpillableMap<>(
+          maxInMemorySizeInBytes,
+          config.getSpillableMapBasePath(),
+          new DefaultSizeEstimator<>(),
+          new DefaultSizeEstimator<>(),
+          config.getCommonConfig().getSpillableDiskMapType(),
+          config.getCommonConfig().isBitCaskDiskMapCompressionEnabled()
+      );
+    } catch (IOException e) {
+      throw new HoodieUpsertException("Failed to initialize HoodieCDCLogger", e);
+    }
+  }
+
+  public void put(HoodieRecord hoodieRecord,
+                  GenericRecord oldRecord,
+                  Option<IndexedRecord> newRecord) {
+    if (cdcEnabled) {
+      String recordKey = hoodieRecord.getRecordKey();
+      GenericData.Record cdcRecord;
+      if (newRecord.isPresent()) {
+        GenericRecord record = (GenericRecord) newRecord.get();
+        if (oldRecord == null) {
+          // inserted cdc record
+          cdcRecord = createCDCRecord(HoodieCDCOperation.INSERT, recordKey,
+              null, record);
+        } else {
+          // updated cdc record
+          cdcRecord = createCDCRecord(HoodieCDCOperation.UPDATE, recordKey,
+              oldRecord, record);
+        }
+      } else {
+        // deleted cdc record
+        cdcRecord = createCDCRecord(HoodieCDCOperation.DELETE, recordKey,
+            oldRecord, null);
+      }
+      cdcData.put(recordKey, new HoodieAvroPayload(Option.of(cdcRecord)));
+    }
+  }
+
+  private GenericData.Record createCDCRecord(HoodieCDCOperation operation,
+                                             String recordKey,
+                                             GenericRecord oldRecord,
+                                             GenericRecord newRecord) {
+    GenericData.Record record;
+    if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE_AFTER)) {
+      record = HoodieCDCUtils.cdcRecord(operation.getValue(), commitTime,
+          removeCommitMetadata(oldRecord), newRecord);
+    } else if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE)) {
+      record = HoodieCDCUtils.cdcRecord(operation.getValue(), recordKey,
+          removeCommitMetadata(oldRecord));
+    } else {
+      record = HoodieCDCUtils.cdcRecord(operation.getValue(), recordKey);
+    }
+    return record;
+  }
+
+  private GenericRecord removeCommitMetadata(GenericRecord record) {
```

Comment:
We can re-use `HoodieAvroUtils.rewriteRecordWithNewSchema`

Reply:
there are a little different: here just want to remove some meta fields, but `rewriteRecordWithNewSchema` will deal each of field in a strict mode. I think `removeCommitMetadata` is more efficient and can be moved to `HoodieAvroUtils`. @alexeykudinkin WDYT ?

Reply:
done.